---
title: "Tidy Transcriptomics For Single-Cell RNA Sequencing Analyses"
author:
  - Stefano Mangiola, Walter and Eliza Hall Institute^[<mangiola.s at wehi.edu.au>]
  - Maria Doyle, Peter MacCallum Cancer Centre^[<maria.doyle at petermac.org>]
output: rmarkdown::html_vignette
bibliography: "`r file.path(system.file(package='bioc2022tidytranscriptomics', 'vignettes'), 'tidytranscriptomics.bib')`"
vignette: >
  %\VignetteIndexEntry{Tidy Transcriptomics For Single-Cell RNA Sequencing Analyses}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Workshop introduction

<p float="left">
<img style="height:100px;" alt="BioC2022" src="https://bioc2022.bioconductor.org/img/carousel/BioC2022.png"/>
<img style="height:100px;" alt="tidybulk" src="https://github.com/Bioconductor/BiocStickers/blob/master/tidybulk/tidybulk.png?raw=true"/>
</p>

## Instructors

*Dr. Stefano Mangiola* is currently a Postdoctoral researcher in the laboratory of Prof. Tony Papenfuss at the Walter and Eliza Hall Institute in Melbourne, Australia. His background spans from biotechnology to bioinformatics and biostatistics. His research focuses on prostate and breast tumour microenvironment, the development of statistical models for the analysis of RNA sequencing data, and data analysis and visualisation interfaces.

*Dr. Maria Doyle* is the Application and Training Specialist for Research Computing at the Peter MacCallum Cancer Centre in Melbourne, Australia. She has a PhD in Molecular Biology and currently works in bioinformatics and data science education and training. She is passionate about supporting researchers, reproducible research, open source and tidy data.

## Description

This tutorial will showcase analysis of single-cell RNA sequencing data following the tidy data paradigm. The tidy data paradigm provides a standard way to organise data values within a dataset, where each variable is a column, each observation is a row, and data is manipulated using an easy-to-understand vocabulary. Most importantly, the data structure remains consistent across manipulation and analysis functions.

This can be achieved with the integration of packages present in the R CRAN and Bioconductor ecosystem, including [tidySingleCellExperiment](https://stemangiola.github.io/tidySingleCellExperiment/), [tidySummarizedExperiment](https://stemangiola.github.io/tidySummarizedExperiment/), [tidybulk](https://stemangiola.github.io/tidybulk/) and [tidyverse](https://www.tidyverse.org/). These packages are part of the tidytranscriptomics suite that introduces a tidy approach to RNA sequencing data representation and analysis. For more information see the [tidy transcriptomics blog](https://stemangiola.github.io/tidytranscriptomics/).

### Pre-requisites

* Basic familiarity with single-cell transcriptomic analyses
* Basic familiarity with tidyverse

## Goals and objectives

* To approach single-cell data representation and analysis through a tidy data paradigm, integrating tidyverse with tidySingleCellExperiment.
* Compare SingleCellExperiment and tidy representation  
* Apply tidy functions to SingleCellExperiment objects  
* Reproduce a real-world case study that showcases the power of tidy single-cell methods

### What you will learn

-   Basic `tidy` operations possible with `tidySingleCellExperiment`
-   The differences between `SingleCellExperiment` representation and `tidy` representation
-   How to interface `SingleCellExperiment` with tidy manipulation and visualisation
-   A real-world case study that will showcase the power of `tidy` single-cell methods compared with base/ad-hoc methods

### What you will *not* learn

-   The molecular technology of single-cell sequencing
-   The fundamentals of single-cell data analysis
-   The fundamentals of tidy data analysis

This workshop will demonstrate a real-world example of using tidy transcriptomics packages to analyse single cell data. This workshop is not a step-by-step introduction in how to perform single-cell analysis. For an overview of single-cell analysis steps performed in a tidy way please see the [ISMB2021 workshop](https://tidytranscriptomics-workshops.github.io/ismb2021_tidytranscriptomics/articles/tidytranscriptomics.html).

## Getting started

### Cloud

Easiest way to run this material. We will use the Orchestra Cloud platform during the BioC2022 workshop.

1. Go to [Orchestra](http://app.orchestra.cancerdatasci.org/).
2. Log in.
3. Find the workshop. In the search box type bioc2022, sort by Created column, and select the most recently created workshop called "BioC2022: Tidy Transcriptomics For Single-Cell RNA Sequencing Analyses" **There are several tidy transcriptomics workshops. Be sure to select the BioC2022 one with the most recent created date**.
4. Click "Launch" (may take a minute or two). 
5. Follow instructions.. **Do not share your personalized URL for the RStudio session, or use the trainers, as only one browser at a time can be connected.**
6. Open `tidytranscriptomics_case_study.Rmd` in `bioc2022_tidytranscriptomcs/vignettes` folder

### Local

We will use the Orchestra Cloud platform during the BioC2022 workshop and this method is available if you want to run the material after the workshop. If you want to install on your own computer, see instructions [here](https://tidytranscriptomics-workshops.github.io/bioc2022_tidytranscriptomics/index.html#workshop-package-installation).

Alternatively, you can view the material at the workshop webpage [here](https://tidytranscriptomics-workshops.github.io/bioc2022_tidytranscriptomics/articles/tidytranscriptomics_case_study.html).

## Slides

*The embedded slides below may take a minute to appear. You can also view or download [here](https://github.com/tidytranscriptomics-workshops/bioc2022_tidytranscriptomics/blob/master/inst/bioc2022_tidytranscriptomics.pdf).*

<iframe 
    src="https://docs.google.com/gview?url=https://raw.githubusercontent.com/tidytranscriptomics-workshops/bioc2022_tidytranscriptomics/master/inst/bioc2022_tidytranscriptomics.pdf&embedded=true" 
    scrolling="yes" 
    style="width:100%; height:600px;" 
    frameborder="0">
</iframe>


# Part 1 Introduction to tidySingleCellExperiment

```{r message = FALSE}
# Load packages
library(SingleCellExperiment)
library(ggplot2)
library(plotly)
library(dplyr)
library(colorspace)
library(dittoSeq)
```

SingleCellExperiment is a very popular analysis toolkit for single cell RNA sequencing data [@butler2018integrating; @stuart2019comprehensive].

Here we load single-cell data in SingleCellExperiment object format. This data is peripheral blood mononuclear cells (PBMCs) from metastatic breast cancer patients.


```{r}
# load single cell RNA sequencing data
sce_obj <- bioc2022tidytranscriptomics::sce_obj

# take a look
sce_obj
```

tidySingleCellExperiment provides a bridge between the SingleCellExperiment single-cell package and the tidyverse [@wickham2019welcome]. It creates an invisible layer that enables viewing the SingleCellExperiment object as a tidyverse tibble, and provides SingleCellExperiment-compatible *dplyr*, *tidyr*, *ggplot* and *plotly* functions.

If we load the *tidySingleCellExperiment* package and then view the single cell data, it now displays as a tibble.

```{r message = FALSE}
library(tidySingleCellExperiment)

sce_obj
```

If we want to revert to the standard SingleCellExperiment view we can do that.

```{r}
options("restore_SingleCellExperiment_show" = TRUE)
sce_obj
```

If we want to revert back to tidy SingleCellExperiment view we can.

```{r}
options("restore_SingleCellExperiment_show" = FALSE)
sce_obj
```

It can be interacted with using [SingleCellExperiment commands](https://bioconductor.org/packages/devel/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html) such as `assays`.

```{r}
assays(sce_obj)
```

We can also interact with our object as we do with any tidyverse tibble.

## Tidyverse commands

We can use tidyverse commands, such as `filter`, `select` and `mutate` to explore the tidySingleCellExperiment object. Some examples are shown below and more can be seen at the tidySingleCellExperiment website [here](https://stemangiola.github.io/tidySingleCellExperiment/articles/introduction.html#tidyverse-commands-1).

We can use `filter` to choose rows, for example, to see just the rows for the cells in G1 cell-cycle stage.

```{r}
sce_obj |> filter(Phase == "G1")
```

We can use `select` to view columns, for example, to see the filename, total cellular RNA abundance and cell phase.

```{r}
sce_obj |> select(file, nCount_RNA, Phase)
```
A couple of things to note with `select`

* As we did not output the .cell column we get a tibble instead of a SingleCellExperiment object and a message to let us know: `tidySingleCellExperiment says: Key columns are missing. A data frame is returned for independent data analysis.` This is ok as it's what we want here when viewing the columns.

* If we use `select` to output the .cell (key) column we will also get any view-only columns returned, such as the UMAP columns generated during the preprocessing.
    ```{r message=FALSE}
sce_obj |> select(.cell, nCount_RNA, Phase)
    ```

We can use `mutate` to create a column. For example, we could create a new `Phase_l` column that contains a lower-case version of `Phase`.

```{r message=FALSE}
sce_obj |>
  mutate(Phase_l = tolower(Phase)) |>
  select(Phase, Phase_l)
```

We can use tidyverse commands to polish an annotation column. We will extract the sample, and group information from the file name column into separate columns.

```{r message=FALSE}
# First take a look at the file column
sce_obj |> select(file)
```

```{r}
# Create column for sample
sce_obj <- sce_obj |>
  # Extract sample
  extract(file, "sample", "../data/.*/([a-zA-Z0-9_-]+)/outs.+", remove = FALSE)

# Take a look
sce_obj |> select(sample, everything())
```

We could use tidyverse `unite` to combine columns, for example to create a new column for sample id combining the sample and patient id (BCB) columns.

```{r message=FALSE}
sce_obj <- sce_obj |> unite("sample_id", sample, BCB, remove = FALSE)

# Take a look
sce_obj |> select(sample_id, sample, BCB)
```


# Part 2 Signature visualisation

## Data pre-processing

The object `sce_obj` we've been using was created as part of a study on breast cancer systemic immune response. Peripheral blood mononuclear cells have been sequenced for RNA at the single-cell level. The steps used to generate the object are summarised below.

-   `scran`, `scater`, and `DropletsUtils` packages have been used to eliminate empty droplets and dead cells. Samples were individually quality checked and cells were filtered for good gene coverage.

-   Variable features were identified using `modelGeneVar`.

-   Read counts were scaled and normalised using logNormCounts from `scuttle`.

-   Data integration was performed using `fastMNN` with default parameters.

-   PCA performed to reduce feature dimensionality.

-   Nearest-neighbor cell networks were calculated using 30 principal components.

-   2 UMAP dimensions were calculated using 30 principal components.

-   Cells with similar transcriptome profiles were grouped into clusters using Louvain clustering from `scran`.

## Analyse custom signature

The researcher analysing this dataset wanted to identify gamma delta T cells using a gene signature from a published paper [@Pizzolato2019]. We'll show how that can be done here.

With tidySingleCellExperiment's `join_features` we can view the counts for genes in the signature as columns joined to our single cell tibble.

```{r}
sce_obj |>
  join_features(c("CD3D", "TRDC", "TRGC1", "TRGC2", "CD8A", "CD8B"), shape = "wide")
```

We can use tidyverse `mutate` to create a column containing the signature score. To generate the score, we scale the sum of the 4 genes, CD3D, TRDC, TRGC1, TRGC2, and subtract the scaled sum of the 2 genes, CD8A and CD8B. `mutate` is powerful in enabling us to perform complex arithmetic operations easily.

```{r}
sce_obj |>
    
  join_features(c("CD3D", "TRDC", "TRGC1", "TRGC2", "CD8A", "CD8B"), shape = "wide") |>
    
  mutate(
    signature_score =
      scales::rescale(CD3D + TRDC + TRGC1 + TRGC2, to = c(0, 1)) -
        scales::rescale(CD8A + CD8B, to = c(0, 1))
  ) |>
    
  select(signature_score, everything())
```

The gamma delta T cells could then be visualised by the signature score using Bioconductor's visualisation functions.

```{r}
sce_obj |>
    
  join_features(
    features = c("CD3D", "TRDC", "TRGC1", "TRGC2", "CD8A", "CD8B"), shape = "wide"
  ) |>
    
  mutate(
    signature_score =
      scales::rescale(CD3D + TRDC + TRGC1 + TRGC2, to = c(0, 1)) -
        scales::rescale(CD8A + CD8B, to = c(0, 1))
  ) |>
    
  scater::plotUMAP(colour_by = "signature_score")
```

The cells could also be visualised using the popular and powerful `ggplot2` package, enabling the researcher to use ggplot functions they were familiar with, and to customise the plot with great flexibility.

```{r}
sce_obj |>
    
  join_features(
    features = c("CD3D", "TRDC", "TRGC1", "TRGC2", "CD8A", "CD8B"), shape = "wide"
  ) |>
    
  mutate(
    signature_score =
      scales::rescale(CD3D + TRDC + TRGC1 + TRGC2, to = c(0, 1)) -
        scales::rescale(CD8A + CD8B, to = c(0, 1))
  ) |>
    
  # plot cells with high score last so they're not obscured by other cells
  arrange(signature_score) |>
    
  ggplot(aes(UMAP_1, UMAP_2, color = signature_score)) +
  geom_point() +
  scale_color_distiller(palette = "Spectral") +
  bioc2022tidytranscriptomics::theme_multipanel
```

For exploratory analyses, we can select the gamma delta T cells, the red cluster on the left with high signature score. We'll filter for cells with a signature score > 0.7.

```{r}
sce_obj_gamma_delta <-
    
  sce_obj |>
    
  join_features(
    features = c("CD3D", "TRDC", "TRGC1", "TRGC2", "CD8A", "CD8B"), shape = "wide"
  ) |>
    
  mutate(
    signature_score =
      scales::rescale(CD3D + TRDC + TRGC1 + TRGC2, to = c(0, 1)) -
        scales::rescale(CD8A + CD8B, to = c(0, 1))
  ) |>
    
	# Proper cluster selection should be used instead (see supplementary material)
  filter(signature_score > 0.7)
```

For comparison, we show the alternative using base R and SingleCellExperiment. Note that the code contains more redundancy and intermediate objects.

```{r eval=FALSE}
counts_positive <-
  assay(sce_obj, "logcounts")[c("CD3D", "TRDC", "TRGC1", "TRGC2"), ] |>
  colSums() |>
  scales::rescale(to = c(0, 1))

counts_negative <-
  assay(sce_obj, "logcounts")[c("CD8A", "CD8B"), ] |>
  colSums() |>
  scales::rescale(to = c(0, 1))

sce_obj$signature_score <- counts_positive - counts_negative

sce_obj_gamma_delta <- sce_obj[, sce_obj$signature_score > 0.7]
```

We can then focus on just these gamma delta T cells and chain Bioconductor and tidyverse commands together to analyse.

```{r warning=FALSE, message=FALSE}
library(batchelor)
library(scater)

sce_obj_gamma_delta <-
    
  sce_obj_gamma_delta |>
    
  # Integrate - using batchelor.
  multiBatchNorm(batch = colData(sce_obj_gamma_delta)$sample) |>
  fastMNN(batch = colData(sce_obj_gamma_delta)$sample) |>
    
  # Join metadata removed by fastMNN - using tidyverse
  left_join(as_tibble(sce_obj_gamma_delta)) |>
    
  # Dimension reduction - using scater
  runUMAP(ncomponents = 2, dimred = "corrected")
```

Visualise gamma delta T cells. As we have used rough threshold we are left with only few cells. Proper cluster selection should be used instead (see supplementary material).

```{r}	
sce_obj_gamma_delta |> plotUMAP()	
```


It is also possible to visualise the cells as a 3D plot using plotly.
The example data used here only contains a few genes, for the sake of time and size in this demonstration, but below is how you could generate the 3 dimensions needed for 3D plot with a full dataset. 

```{r eval = FALSE}
single_cell_object |>
  RunUMAP(dims = 1:30, n.components = 3L, spread = 0.5, min.dist = 0.01, n.neighbors = 10L)
```

We'll demonstrate creating a 3D plot using some data that has 3 UMAP dimensions. This is a fantastic way to visualise both reduced dimensions and metadata in the same representation. 

```{r umap plot 2, message = FALSE, warning = FALSE}
pbmc <- bioc2022tidytranscriptomics::sce_obj_UMAP3

pbmc |>
  plot_ly(
    x = ~`UMAP_1`,
    y = ~`UMAP_2`,
    z = ~`UMAP_3`,
    color = ~cell_type,
    colors = dittoSeq::dittoColors()
  ) %>%
  add_markers(size = I(1))
```

## Exercises

Using the `sce_obj`:

1. What proportion of all cells are gamma-delta T cells? Use signature_score > 0.7 to identify gamma-delta T cells.

2. There is a cluster of cells characterised by a low RNA output (nCount_RNA < 100). Identify the cell composition (cell_type) of that cluster.

# Part 3 Pseudobulk analyses

Next we want to identify genes whose transcription is affected by treatment in this dataset, comparing treated and untreated patients. We can do this with pseudobulk analysis. We aggregate cell-wise transcript abundance into pseudobulk samples and can then perform hypothesis testing using the very well established bulk RNA sequencing tools. For example, we can use edgeR in tidybulk to perform differential expression testing. For more details on pseudobulk analysis see [here](https://hbctraining.github.io/scRNA-seq/lessons/pseudobulk_DESeq2_scrnaseq.html).

We want to do it for each cell type and the tidy transcriptomics ecosystem makes this very easy. 


## Create pseudobulk samples

To create pseudobulk samples from the single cell samples, we will use a helper function called `aggregate_cells`, available in this workshop package. This function will combine the single cells into a group for each cell type for each sample.

```{r warning=FALSE, message=FALSE, echo=FALSE}
library(glue)
library(tidyr)
library(purrr)
library(patchwork)

# bulk RNA-seq libraries
library(tidybulk)
library(tidySummarizedExperiment)
```

```{r}
pseudo_bulk <-
  sce_obj |>
  bioc2022tidytranscriptomics::aggregate_cells(c(sample, cell_type), assays = "counts")

pseudo_bulk
```

## Tidybulk and tidySummarizedExperiment

With `tidySummarizedExperiment` and `tidybulk` it is easy to split the data into groups and perform analyses on each without needing to create separate objects.

```{r, echo=FALSE, out.width = "800px"}
knitr::include_graphics("../man/figures/new_SE_usage-01.png")
```

We use tidyverse `nest` to group the data. The command below will create a tibble containing a column with a SummarizedExperiment object for each cell type. `nest` is similar to tidyverse `group_by`, except with `nest` each group is stored in a single row, and can be a complex object such as a plot or SummarizedExperiment.

```{r}
pseudo_bulk_nested <- 
  pseudo_bulk |>
  nest(grouped_summarized_experiment = -cell_type)

pseudo_bulk_nested
```

To explore the grouping, we can use tidyverse `slice` to choose a row (cell_type) and `pull` to extract the values from a column. If we pull the data column we can view the SummarizedExperiment object. 

```{r}
pseudo_bulk_nested |>
  slice(1) |>
  pull(grouped_summarized_experiment)
```

We can then identify differentially expressed genes for each cell type for our condition of interest, treated versus untreated patients. We use tidyverse `map` to apply differential expression functions to each cell type group in the nested data. The result columns will be added to the SummarizedExperiment objects.

```{r message=FALSE, warning=FALSE}
# Differential transcription abundance
pseudo_bulk <-
    
  pseudo_bulk_nested |>
    
  # map accepts a data column (.x) and a function. It applies the function to each element of the column.
  mutate(grouped_summarized_experiment = map(
    grouped_summarized_experiment,
    ~ .x |>
        
      # Removing genes with low expression
      identify_abundant(factor_of_interest = treatment) |>
        
      # Scaling counts for sequencing depth	
      scale_abundance(method="TMMwsp") |>
      
      # Testing for differential expression using edgeR quasi likelihood	
      test_differential_abundance(~treatment, method="edgeR_quasi_likelihood")
  ))
```

The output is again a tibble containing a SummarizedExperiment object for each cell type.

```{r}
pseudo_bulk
```

If we pull out the SummarizedExperiment object for the first cell type, as before, we can see it now has columns containing the differential expression results (e.g. logFC, PValue).

```{r}
pseudo_bulk |>
  slice(1) |>
  pull(grouped_summarized_experiment)
```

It is useful to create plots for significant genes to visualise the transcriptional abundance in the groups being compared (treated and untreated). We can do this for each cell type without needing to create multiple objects. 

```{r message = FALSE}
pseudo_bulk <-
    
  pseudo_bulk |>
    
  # Filter out significant
  # using a high FDR value as this is toy data
  mutate(grouped_summarized_experiment = map(
  	grouped_summarized_experiment, 
  	~ filter(.x, FDR < 0.5)
  )) |>
    
  # Filter cell types with no differential abundant gene-transcripts
  # map_int is map that returns integer instead of list
  filter(map_int(grouped_summarized_experiment, ~ nrow(.x)) > 0) |>

  # Plot significant genes for each cell type
  # map2 is map that accepts 2 input columns (.x, .y) and a function
  mutate(plot = map2(
    grouped_summarized_experiment, cell_type,
    ~ .x |>
      ggplot(aes(treatment, counts_scaled + 1)) +
      geom_boxplot(aes(fill = treatment)) +
      geom_jitter() +
      scale_y_log10() +
      facet_wrap(~.feature) +
      ggtitle(.y)
  ))
```

The output is a nested table with a column containing a plot for each cell type.

```{r}
pseudo_bulk
```

We'll use slice and pull again to have a look at one of the plots.

```{r}
pseudo_bulk |>
  slice(1) |>
  pull(plot)
```

We can extract all plots and plot with `wrap_plots` from the patchwork package. 

```{r, fig.height = 8}
pseudo_bulk |>
  pull(plot) |>
  wrap_plots() &
  bioc2022tidytranscriptomics::theme_multipanel
```

# Feedback

Thank you for attending this workshop. We hope it was an informative session for you. We would be grateful if you could help us by taking a few moments to provide your valuable feedback in the short form below. Your feedback will provide us with an opportunity to further improve the workshop. All the results are anonymous. 

[Feedback Form Link](https://forms.gle/rVzZo6k8QfYW2jJx6) 

**Session Information**

```{r}
sessionInfo()
```

**References**
